### ISSUE TITLE
UI Framework

### ISSUE THREAD

**Issue by Xaeroxe** [amethyst/amethyst#365]
At this moment I feel our most urgent need is for a proper UI framework.  We have a few issues blocking this currently, please check these off as they are corrected:

- [x] This PR should be merged before we make changes to the renderer otherwise https://github.com/amethyst/amethyst/pull/317
- [x] Flipped textures in OpenGL.  Fixed by https://github.com/amethyst/amethyst/pull/377
- [x] No transparency https://github.com/amethyst/amethyst/issues/357
- [x] Renderer does not support resizing https://github.com/amethyst/amethyst/pull/402

@Aceeri has thus far been working on https://github.com/Aceeri/citrine which will be helpful to this.

EDIT: Update the Design thus far to reflect ongoing conversation. 2017-09-19

Design thus far: Implement a system with three fundamental UI building blocks:

- Text
- Image
- Canvas

Where Canvas is simply a rectangle with more UI building blocks in it.  Each building block will have an absolute position on screen, a width, and a height.  This will permit maximum flexibility in the early stages.  From here we can easily create buttons, scroll bars, and in theory whatever else we need.

**Comment by jojolepro** [comment]
UI is the most urgent functionality, as well as at least a basic physic engine.

**Comment by torkleyy** [comment]
@Xaeroxe Didn't we agree on using absolute coordinate for the first iteration, and adding support for things like anchors, relative positioning, ... with Layouts?

**Comment by Xaeroxe** [comment]
Sorry @torkleyy additional conversation was had on gitter and I haven't updated this issue with it yet, I'll do that soon.

**Comment by Xaeroxe** [comment]
To summarize: yes we're going to do what you suggest.

**Comment by jojolepro** [comment]
Poke about using html + css. Would allow to reuse existing pages and libraries (bootstrap? :D )
Would also allow a better way to handle resizes, like mobile-compatible websites do. The only drawback I see is because of how complicated it can be to get a web rendering engine in. Also the rust-html-css animations might be non-trivial. I haven't really checked if there is already existing libraries that are doing that. And, more importantly, we won't reinvent the wheel once again by recreating a whole UI framework with basic things like text, buttons, links, etc...

**Comment by minecrawler** [comment]
no one likes bootstrap ;P

Well, Servo can render html+css and afaik has a JS engine. It exposes, amongst others, the CEF API, so I am pretty sure that it is possible to get some JS<->Rust interaction. The [Servoshell](https://github.com/paulrouget/servoshell) is a very nice resource which demonstrates how to use Servo. It would also be my preferred GUI renderer. One of the downsides I see is that it might be hard to integrate.

**Comment by Xaeroxe** [comment]
I'm kind of back and forth on this because while HTML + CSS are battle proven they've also been shown to be largely successful today because of developers using things in ways that they weren't exactly meant to be used. So I kind of want to start from scratch, let's try again and see if we can do better this time bearing in mind the mistakes of the past.

**Comment by minecrawler** [comment]
@Xaeroxe why are you against using standards? Developers are already familiar with HTML+CSS+JS, not with AmethystXML, AmethystSS and AmethystScript. Standards also integrate nicely with all the millions of tools already available. I think, saying that we should try to make a revolution here is contra-productive.

However, I also think that implementing Servo or similar at this point would keep us from doing more important things, so implementing a more simple GUI with your ideas is a better way to go for now.

**Comment by anderejd** [comment]
Anyone tried using conrod in combination with amethyst? Or some other existing rust GUI framework?

Link to conrod: http://docs.piston.rs/conrod/conrod/guide/chapter_1/index.html

**Comment by Xaeroxe** [comment]
@andrejd yes we've considered using it, however we wanted to use something that could integrate into specs so we decided to make our own.

**Comment by anderejd** [comment]
@Xaeroxe Sounds like an interesting aproach, that could allow really wild GUIs. Any documentation on that discussion available?

**Comment by Xaeroxe** [comment]
Unfortunately no, the conversation has been disparate and scattered, this issue is the best unification of the conversation that we have.

**Comment by jojolepro** [comment]
Didn't we had a 50-60 post long discussion on another issue not too long ago?

**Comment by Xaeroxe** [comment]
If we did I can't find it :/

**Comment by Aceeri** [comment]
Messing with servo to patch in HTML and CSS seems like more work to me honestly. How would we communicate with it? Javascript?

**Comment by minecrawler** [comment]
> Javascript?

In case of Servo, probably yes.

> Id rather integrate with specs directly and make amethyst more seamless by having most things held together by specs systems than just patching in "somewhat working" tools.

Which does not mean that we shouldn't use standards. I honestly don't care about the implementation. I care about being able to use it.

**Comment by omni-viral** [comment]
Most tools for working with html and css imply using js as well. Communications is also supposed to go via js. And I don't think it is the best choice.

**Comment by Noxivs** [comment]
It could be interesting to see if [WebRender](https://github.com/servo/webrender) matches with what you expect for your UI Framework. Note that [Pathfinder 2](https://github.com/pcwalton/pathfinder) will be added to Webrender.

Some rust developers tried to use WR for UI already :

https://github.com/Thinkofname/stylish 
https://github.com/trishume/quickdown

**Comment by erlend-sh** [comment]
> we've considered using it, however we wanted to use something that could integrate into specs so we decided to make our own.

Did anyone ever contact @mitchmindtree to ask if this might be a good path forward for Conrod? @bvssvni did briefly poke at the possibility of using an ECS: https://github.com/PistonDevelopers/conrod/issues/11

**Comment by torkleyy** [comment]
I would be in favor of such a solution @erlend-sh. Using HTML & CSS for game UI seems too heavy and unpractical to me, but I must also admit that my HTML knowledge isn't very good. Most important things really are 1) ECS integration and 2) performance.

**Comment by minecrawler** [comment]
@torkleyy 

> Using HTML & CSS for game UI seems too heavy and unpractical to me,

It's a common misbelief, actually. There are a lot of web-renderers for game UI for C++, and they are even used in big commercial games. One example is [CoherentUI](http://coherent-labs.com/), which is used in [Guild Wars 2](https://www.guildwars2.com/en/) among others (check their partner list). Others include [Awesomium](http://wiki.awesomium.com/) and [Sciter](https://github.com/sciter-sdk/rust-sciter). They are used for the GUI and to display websites in-game (useful for an in-game-store for example).

Today's web renderers are very powerful (yes, they used to be slow!), because the world is moving more and more towards web apps. People are work on [WebGL](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API) and [WASM](https://developer.mozilla.org/en-US/docs/WebAssembly), which bring GPU accelerated graphics and typed and compiled languages (Rust <3) to the browser in order to create desktop applications _and performance-hungry games_ as a service! So, while you might have even more prejudice against the `as a service` move, performance is imperative for the web of today. 

Now, the question, which remains, is: How to put that thing into an ECS? Imho, it's all about API. So, I think the key is to use JS as glue and provide a Rust-API to the sandbox, which enables JS to create new entities, register components on them and modify them during a round of JS-task-execution (which can be threaded). This is the hard part and getting it right will probably take up quite a bit of time. It's completely out of scope while we cannot even create a basic game which is a bit larger than those example-demos with Amethyst :D

Since we plan to separate the parts of Amethyst, it should not be a lot of trouble to provide more than one alternative (for example @Xaeroxe's ECS-GUI, Conrod and some HTML-implementation), as long as someone dedicates the time to maintain it. I'd be happy with going for that approach and having the ECS-GUI as the default and described in basic tutorials, but also leave the possibility to go for a more heavy approach when I need it.

As for WebRender, it's a part of Servo, so no benefit over that as far as I can see

**Comment by Rhuagh** [comment]
GW2 only use coherent for the ui parts that are fetched from the Web, like the trading post, not the whole ui.

**Comment by omni-viral** [comment]
I think I'm going to try to integrate conrod. Just to have at least some UI until we have better solution.

**Comment by torkleyy** [comment]
@minecrawler I fully agree with your last point. One can just write a Specs specific GUI library. There's simply no need to settle on one solution, at least not now. But personally I still don't like writing my GUI in HTML. Maybe I do in the future, who knows.

**Comment by Rhuagh** [comment]
I don't wanna write my UI in HTML either, I've written a ton of HTML/CSS for my job, and it's absolutely awful.

**Comment by OvermindDL1** [comment]
I also write html/css/javscript/(+backend stuff) for my job, ew...

I'm personally quite partial to the [(fake) immediate mode](https://github.com/ocornut/imgui) (that is not the only one out, just the one I've been using) GUI systems in C++.  It looks entirely intermediate but it builds up a virtual gui backend (in reality it is even more optimized than that, only things that actually have state store anything at all, and for example even a button just has a few bytes of state) and just diff's from call to call.  It is highly efficient (it outperforms most other game gui systems by quite a wide margin), easy to change, easy to put effects with, and is exceedingly more suitable for games since you have a significant amount more control and you don't have to do weird binding and non-local function calls to interact with it (little to no virtual calls whoo!).

**Comment by Rhuagh** [comment]
I think the ECS version we are building is something along those lines, we won't have any "logic" in the UI parts, that's up to the user to drive, we will just do the layouting, rendering and basic input handling (active/hot etc).

**Comment by Xaeroxe** [comment]
![image](https://i.ytimg.com/vi/euAVJaxdd-U/hqdefault.jpg)

**Comment by remram44** [comment]
PR for reference: #441. Great work!

**Comment by oraknabo** [comment]
I'm only just getting started with Rust and Amethyst, but I think the lack of a comprehensive UI system is the only thing that's been stopping me from committing to developing a full game with them. I'm glad to see @Xaeroxe's progress on this but would like to know more about plans for future development.

I think immediate mode works best for dev tools & editors, but for the actual game GUI, I'd like to see something more visually extensible. I like the idea of using specs with a canvas-style system. I don't know what BitSquid's UI system is like, but I like the newer one in Unity that uses their ECS and would love to see something like that, especially the way you can define scaling borders around panels & buttons.

I do a lot of HTML/CSS/JS frontend work in my day job and I agree with @minecrawler that there's a lot of potential to the idea of an alternative system like Awesomium based on Servo's WebRender. EA's been doing their UI this way in games like Sims & Sim City for a while now. Here are the slides from a talk about it from a couple of years ago: [https://www.gdcvault.com/play/1022055/How-to-Implement-AAA-Game](url) for anyone interested. 

It would be interesting to try something HTML-based built around wasm or with a more rigorous language than JS like Elm, but I'm sure there would be limitations to both that make it much harder to pull off. I understand that you lose some performance interpreting JS & parsing the DOM while embedding a browser in your game, but it seems like you could gain a decent about flexibility in trade.

Is the plan for the UI to use an events system for communicating with other entities/components in the game world? Will that be part of Citrine?

**Comment by oraknabo** [comment]
Sorry, just realized this is a closed issue after writing all that. I'll check back later for when a new issue gets put up to discuss future development on the UI systems.

**Comment by Aceeri** [comment]
@oraknabo Feel free to open an issue here or on citrine's issue tracker, we still have a ways to go with UI stuff.

**Comment by i-schuetz** [comment]
Any news about GUI development? Citrine (linked above) hasn't had activity in the last 3 years.

**Comment by erlend-sh** [comment]
If native UI system of Amethyst doesn't suffice for you, you might wanna check out the Iced integration by @aliscode: https://github.com/AlisCode/amethyst_iced
